app/src/androidTest/java/ch/epfl/polybazaar/ExampleInstrumentedTest.java

```diff
@@ -0,0 +1,119 @@
+        user.put("last", "Lovelace");
+        user.put("born", 1815);
+
+        // Add a new document with a generated ID
+        db.collection("users")
+                .add(user)
+                .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
+                    @Override
+                    public void onSuccess(DocumentReference documentReference) {
+                        assertEquals(1, 1);
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        assertEquals(1, 2);
+                    }
+                });
+
+        // Add a new document with a generated ID
+        db.collection("unknown")
+                .add(user)
+                .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
+                    @Override
+                    public void onSuccess(DocumentReference documentReference) {
+                        assertEquals(1, 2);
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        assertEquals(1, 1);
+                    }
+                });
+
+        final DocumentReference docRef = db.collection("users").document("user");
+                docRef.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
+                    @Override
+                    public void onComplete(@NonNull Task<DocumentSnapshot> task) {
+                        if (task.isSuccessful()) {
+                            DocumentSnapshot document = task.getResult();
+                            if (document.exists()) {
+                                assertEquals(1, 1);
+                            } else {
+                                assertEquals(1, 2);
+                            }
+                        } else {
+                            assertEquals(1, 2);
+                        }
+                    }
+                });
+        docRef.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
+                if (task.isSuccessful()) {
+                    DocumentSnapshot document = task.getResult();
+                    boolean works = true;
+                    if (document.contains("first") &&
+                    document.contains("last") &&
+                    document.contains("born") &&
+                    document.contains("first") &&
+                    document.get("last").equals("Lovelace") &&
+                    document.get("first").equals("Ada") &&
+                    document.get("born").equals(1815)) assertEquals(1, 1);
+                    else assertEquals(1, 2);
```

Comment:
The tests do not seem to work, for example on line 81 I was able to change `assertEquals(1, 1)` to `assertEquals(1, 2)` and the tests still pass. This is because methods like `onSuccess` are asynchronous and are not necessarily executed before the test ends. This [link](https://firebase.google.com/docs/rules/unit-tests#firestore) might give insight on how to write unit tests for Firebase

Reply:
I noticed that too. Since those tests are quite useless at this stage, I think I'll just get rid of them. Thanks for the link, that'll surely be useful very soon.